readthedocs/rtd_tests/tests/test_backend.py

```diff
@@ -378,6 +389,244 @@ def test_fetch_clean_tags_and_branches(self, checkout_path):
+        """
+        repo_path = self.project.repo
+        branches = [
+            "develop",
+            "master",
+            "2.0.X",
+            "release/2.0.0",
+            "release/foo/bar",
+        ]
+        for branch in branches:
+            create_git_branch(repo_path, branch)
+
+        for branch in branches:
+            # Create a repo that we want to clone and fetch a specific branch for
+            repo = self.project.vcs_repo(
+                environment=self.build_environment,
+                version_identifier=branch,
+                version_type=BRANCH,
+            )
+            # Because current behavior is to reuse already cloned destinations, we should
+            # clear the working dir instead of reusing it.
+            repo.make_clean_working_dir()
+
+            repo.update()
+
+            self.assertEqual(
+                set([branch, "master"]),
+                {branch.verbose_name for branch in repo.branches},
+            )
+
+    def test_git_branches_unicode(self):
+        """Test to verify that we can clone+fetch a unicode branch name."""
+
+        # Add a branch to the repo.
+        # Note: It's assumed that the source repo is re-created in setUp()
+        create_git_branch(self.project.repo, "release-ünîø∂é")
+
+        repo = self.project.vcs_repo(
+            environment=self.build_environment,
+            version_identifier="release-ünîø∂é",
+            version_type=BRANCH,
+        )
+        repo.update()
+
+        # Note here that the original default branch 'master' got created during the clone
+        self.assertEqual(
+            set(["release-ünîø∂é", "master"]),
+            {branch.verbose_name for branch in repo.branches},
+        )
+
+    def test_update_without_branch_name(self):
+        """
+        Test that we get expected default branch when not specifying a branch name.
+
+        Covers:
+        * Manually imported projects +
+        * Automatically imported projects that had their default branch name deliberately removed
+          during import wizard (and before the first branch sync from webhook)
+        """
+        repo_path = self.project.repo
+        create_git_branch(repo_path, "develop")
+
+        repo = self.project.vcs_repo(
+            environment=self.build_environment,
+            version_type=BRANCH,
+        )
+        repo.update()
+
+        # Check that the README file from the fixture exists
+        self.assertTrue(os.path.isfile(os.path.join(repo.working_dir, "README")))
+
+        # Check that "only-on-default-branch" exists (it doesn't exist on other branches)
+        self.assertTrue(
+            os.path.exists(os.path.join(repo.working_dir, "only-on-default-branch"))
+        )
+
+        # Check that we don't for instance have foobar
+        self.assertFalse(os.path.isfile(os.path.join(repo.working_dir, "foobar")))
+
+    def test_special_tag_stable(self):
```

Comment:
This test is also new. Have to do some advanced stuff if we are going to verify the part with commit hashes.. but I'm inclined to think it's good to do that.

Reply:
(we don't know the commit hashes for the repository in the fixtures, so we'd have to scrape those in order to use them in a test)

Reply:
("stable" Versions will check out the commit hash directly through `version_identifier`, so it'd be nice to actually test that this part works, no matter if the commit is part of the default branch or some other branch)

Reply:
Fixed in 41e1f09df